go-controller/pkg/network-attach-def-controller/network_attach_def_controller.go

```diff
@@ -280,3 +323,50 @@ func nadNeedsUpdate(oldNAD, newNAD *nettypes.NetworkAttachmentDefinition) bool {
 
 	return !reflect.DeepEqual(oldNAD.Spec, newNAD.Spec)
 }
+
+func (nadController *NetAttachDefinitionController) GetActiveNetworkForNamespace(namespace string) (util.NetInfo, error) {
+	if !util.IsNetworkSegmentationSupportEnabled() {
+		return &util.DefaultNetInfo{}, nil
+	}
+	nadController.RLock()
+	defer nadController.RUnlock()
+	primaryNAD := nadController.primaryNADs[namespace]
+	if primaryNAD != "" {
+		// we have a primary NAD, get the network
+		netName := nadController.nads[primaryNAD]
+		if netName == "" {
+			// this should never happen where we have a nad keyed in the primaryNADs
+			// map, but it doesn't exist in the nads map
+			panic("NAD Controller broken consistency between primary NADs and cached NADs")
+		}
+		network := nadController.networkManager.getNetwork(netName)
+		networkNADs := network.GetNADs()
+		filteredNADs := make([]string, 0, len(networkNADs))
+		for _, nadName := range networkNADs {
+			nadNs, _, err := cache.SplitMetaNamespaceKey(nadName)
+			if err != nil {
+				klog.Errorf("Cannot parse NAD reference: %q on network info: %q", nadName, netName)
+				continue
+			}
+			if nadNs == namespace {
+				filteredNADs = append(filteredNADs, nadName)
+			}
+		}
+		n := util.CopyNetInfo(network)
+		n.SetNADs(filteredNADs...)
```

Comment:
If we already had the NADs per namespace in NetInfo, which is an information we already come by naturally in sync method, we wouldn't have to do this search every time GetActiveNetworkForNamespace is called

Reply:
I am also wondering...should we disallow multiple NADs per namespace on a primary network?

So there could be multiple NADs per namespace for a primary network as long as only one of them is flagged as primary. That would be an interesting use case if, let's say, selected pods want to customize how they attach to the primary network, for example use sr-iov. The one flagged as primary would be used by default. The others would be used through the network selection element annotation.

We currently disallow the network selection element annotation for primary networks so there's actually no use for that. However we can't prevent the mere existence of multiple NADs per namespace for a primary network. So we either have a way to tell around which one of them is really primary or we don't make anyone aware of those NADs since they have no use for us.

Reply:
To disallow, we would just have to check against the current network instead of the nad network:
```
if currentNetwork != nil && currentNetwork.IsPrimaryNetwork() && primaryNAD != "" && primaryNAD != key
```

Then it his method you would just have to do
```
n.SetNADs(primaryNAD)
```

Reply:
but we have a way to really tell which is the active one, since we have the primaryNADs map? Why don't i just use that?